observIQ/bindplane-op
# feat(cli): Support exporting resources with get command

### Proposed Change


Added "export" support to the following resources
- sources
- processors
- destinations
- configurations

When the `--export` flag is passed to the `get` command, resources will be returned without select metadata such as `id`, `hash`, `date modified`, `status`, etc. Additionally, version suffix is removed from the type, such as `file:3` --> `file`.

By exporting resources, they will be in a fresh state suitable for import into a new system.

### Testing

You can test with the following script. 

1. Create one or more destinations
2. Create a configuration which uses those destinations
3. Rollout the configuration
4. Make changes to destinations and sources
5. Rollout the configuration

At this point, you have a configuration and destinations with several versions.

Export all resources:
```bash
bindplane get source -o yaml --export > source.yaml
bindplane get processor -o yaml --export > processor.yaml
bindplane get destination -o yaml --export > destination.yaml
bindplane get configuration -o yaml --export > configuration.yaml
```

You can inspect the output files. Note that they do not include version information, or generated IDs and hashes.

Delete your configurations and destinations and then import the resources from your yaml files. This will simulate a migration from one system to another.

```
bindplane apply -f source.yaml
bindplane apply -f processor.yaml
bindplane apply -f destination.yaml
bindplane apply -f configuration.yaml
```

Running a `bindplane get config <name> -o yaml` should should your configuration as version 1. You should see similar results with destinations.


I tested with Terraform, creating > 10 sources, > 10 processors, > 10 destinations and around 100 configurations. I was able to export everything and import to a fresh system.

##### Checklist
